mmdet/models/dense_heads/solo_head.py

```diff
@@ -0,0 +1,1039 @@
+                ins_pred_y = F.interpolate(
+                    ins_pred_y.sigmoid(), size=upsampled_size, mode='bilinear')
+                cate_pred = points_nms(
+                    cate_pred.sigmoid(), kernel=2).permute(0, 2, 3, 1)
+            ins_pred_x_maps.append(ins_pred_x)
+            ins_pred_y_maps.append(ins_pred_y)
+            cate_pred_maps.append(cate_pred)
+        return ins_pred_x_maps, ins_pred_y_maps, cate_pred_maps
+
+    def loss(self,
+             ins_preds_x,
+             ins_preds_y,
+             cate_preds,
+             gt_labels,
+             gt_masks,
+             img_metas,
+             gt_bboxes=None,
+             gt_bboxes_ignore=None,
+             **kwargs):
+        featmap_sizes = [featmap.size()[-2:] for featmap in ins_preds_x]
+
+        ins_label_list, cate_label_list, \
+            ins_ind_label_list, ins_ind_label_list_xy = \
+            multi_apply(self.solo_target_single,
+                        gt_bboxes,
+                        gt_labels,
+                        gt_masks,
+                        featmap_sizes=featmap_sizes)
+
+        # ins
+        ins_labels = [[] for _ in range(len(ins_preds_x))]
+        ins_preds_x_final = [[] for _ in range(len(ins_preds_x))]
+        ins_preds_y_final = [[] for _ in range(len(ins_preds_x))]
+        cate_labels = [[] for _ in range(len(ins_preds_x))]
+        for i in range(len(ins_label_list)):
+            assert len(ins_preds_x) == len(ins_label_list[i])
+            for j in range(len(ins_label_list[i])):
+                ins_labels[j].append(
+                    ins_label_list[i][j][ins_ind_label_list[i][j], ...])
+                ins_preds_x_final[j].append(
+                    ins_preds_x[j][i, ins_ind_label_list_xy[i][j][:, 1], ...])
+                ins_preds_y_final[j].append(
+                    ins_preds_y[j][i, ins_ind_label_list_xy[i][j][:, 0], ...])
+                cate_labels[j].append(cate_label_list[i][j].flatten())
+
+        cate_pred_temp = []
+        for i in range(len(ins_labels)):
+            ins_labels[i] = torch.cat(ins_labels[i], dim=0)
+            ins_preds_x_final[i] = torch.cat(ins_preds_x_final[i], dim=0)
+            ins_preds_y_final[i] = torch.cat(ins_preds_y_final[i], dim=0)
+            cate_labels[i] = torch.cat(cate_labels[i], dim=0)
+            cate_pred_temp.append(cate_preds[i].permute(0, 2, 3, 1).reshape(
+                -1, self.cate_out_channels))
+        cate_preds = cate_pred_temp
+
+        num_ins = 0.
+        # dice loss
+        loss_mask = []
+        for pred_x, pred_y, target in \
+                zip(ins_preds_x_final, ins_preds_y_final, ins_labels):
+            mask_n = pred_x.size(0)
+            if mask_n == 0:
+                # make sure can get grad
+                loss_mask.append((pred_x.sum() + pred_y.sum()).unsqueeze(0))
+                continue
+            num_ins += mask_n
+            loss_mask.append(self.loss_mask((pred_x, pred_y), target))
```

Comment:
Need to  make dice loss more flexible.
Now is feeding A & B to dice loss and using sigmoid(A)+sigmoid(B) in Dice loss.